File: src/os/syscalls.ts
/**
 * JSOS System Call Interface
 *
 * Provides a clean, typed interface for system calls between the TypeScript OS
 * layer and the underlying kernel. This replaces direct kernel API calls with
 * proper system call numbers, parameter validation, and return value handling.
 */

import { KernelAPI } from './kernel.js';

declare var kernel: KernelAPI;

// System call numbers (must match kernel implementation)
export const enum SyscallNumber {
  // Process management
  FORK = 1,
  EXEC = 2,
  EXIT = 3,
  WAIT = 4,
  GETPID = 5,
  GETPPID = 6,
  KILL = 7,

  // Memory management
  BRK = 8,
  MMAP = 9,
  MUNMAP = 10,

  // File I/O
  OPEN = 11,
  CLOSE = 12,
  READ = 13,
  WRITE = 14,
  LSEEK = 15,
  STAT = 16,
  FSTAT = 17,

  // Directory operations
  CHDIR = 18,
  GETCWD = 19,
  MKDIR = 20,
  RMDIR = 21,
  LINK = 22,
  UNLINK = 23,

  // Time
  TIME = 24,
  STIME = 25,

  // System info
  UNAME = 26,
  GETUID = 27,
  GETGID = 28,

  // Signals
  SIGNAL = 29,
  SIGACTION = 30,
  SIGPROCMASK = 31,

  // IPC
  PIPE = 32,
  MSGGET = 33,
  MSGSND = 34,
  MSGRCV = 35,

  // Network (future)
  SOCKET = 40,
  BIND = 41,
  LISTEN = 42,
  ACCEPT = 43,
  CONNECT = 44,
  SEND = 45,
  RECV = 46
}

export interface SyscallResult<T = any> {
  success: boolean;
  value?: T;
  error?: string;
  errno?: number;
}

// Error codes (POSIX-style)
export const enum Errno {
  EPERM = 1,      // Operation not permitted
  ENOENT = 2,     // No such file or directory
  ESRCH = 3,      // No such process
  EINTR = 4,      // Interrupted system call
  EIO = 5,        // I/O error
  ENXIO = 6,      // No such device or address
  E2BIG = 7,      // Argument list too long
  ENOEXEC = 8,    // Exec format error
  EBADF = 9,      // Bad file descriptor
  ECHILD = 10,    // No child processes
  EAGAIN = 11,    // Try again
  ENOMEM = 12,    // Out of memory
  EACCES = 13,    // Permission denied
  EFAULT = 14,    // Bad address
  ENOTBLK = 15,   // Block device required
  EBUSY = 16,     // Device or resource busy
  EEXIST = 17,    // File exists
  EXDEV = 18,     // Cross-device link
  ENODEV = 19,    // No such device
  ENOTDIR = 20,   // Not a directory
  EISDIR = 21,    // Is a directory
  EINVAL = 22,    // Invalid argument
  ENFILE = 23,    // File table overflow
  EMFILE = 24,    // Too many open files
  ENOTTY = 25,    // Not a typewriter
  ETXTBSY = 26,   // Text file busy
  EFBIG = 27,     // File too large
  ENOSPC = 28,    // No space left on device
  ESPIPE = 29,    // Illegal seek
  EROFS = 30,     // Read-only file system
  EMLINK = 31,    // Too many links
  EPIPE = 32,     // Broken pipe
  EDOM = 33,      // Math argument out of domain
  ERANGE = 34     // Math result not representable
}

export class SystemCallInterface {
  private nextFd = 3; // 0=stdin, 1=stdout, 2=stderr

  /**
   * Make a system call to the kernel
   * This is a placeholder - in a real implementation, this would use
   * architecture-specific syscall instructions
   */
  private syscall(num: SyscallNumber, ...args: any[]): SyscallResult {
    try {
      // For now, we'll simulate syscalls by calling kernel functions
      // In a real OS, this would be: syscall(num, arg1, arg2, ...)
      switch (num) {
        case SyscallNumber.GETPID:
          return { success: true, value: 1 }; // Mock PID

        case SyscallNumber.TIME:
          return { success: true, value: Math.floor(Date.now() / 1000) };

        case SyscallNumber.BRK:
          // Mock memory allocation
          return { success: true, value: 0x100000 };

        default:
          return { success: false, error: 'Not implemented', errno: Errno.ENOSYS };
      }
    } catch (e) {
      return { success: false, error: String(e), errno: Errno.EFAULT };
    }
  }

  // Process Management System Calls
  fork(): SyscallResult<number> {
    return this.syscall(SyscallNumber.FORK);
  }

  exec(path: string, args: string[]): SyscallResult<never> {
    return this.syscall(SyscallNumber.EXEC, path, args);
  }

  exit(status: number): void {
    this.syscall(SyscallNumber.EXIT, status);
  }

  wait(pid?: number): SyscallResult<{ pid: number; status: number }> {
    return this.syscall(SyscallNumber.WAIT, pid);
  }

  getpid(): SyscallResult<number> {
    return this.syscall(SyscallNumber.GETPID);
  }

  getppid(): SyscallResult<number> {
    return this.syscall(SyscallNumber.GETPPID);
  }

  kill(pid: number, sig: number): SyscallResult<void> {
    return this.syscall(SyscallNumber.KILL, pid, sig);
  }

  // Memory Management System Calls
  brk(addr?: number): SyscallResult<number> {
    return this.syscall(SyscallNumber.BRK, addr);
  }

  mmap(addr: number, length: number, prot: number, flags: number, fd: number, offset: number): SyscallResult<number> {
    return this.syscall(SyscallNumber.MMAP, addr, length, prot, flags, fd, offset);
  }

  munmap(addr: number, length: number): SyscallResult<void> {
    return this.syscall(SyscallNumber.MUNMAP, addr, length);
  }

  // File I/O System Calls
  open(pathname: string, flags: number, mode?: number): SyscallResult<number> {
    const result = this.syscall(SyscallNumber.OPEN, pathname, flags, mode);
    if (result.success && typeof result.value === 'number') {
      return { success: true, value: this.nextFd++ };
    }
    return result;
  }

  close(fd: number): SyscallResult<void> {
    return this.syscall(SyscallNumber.CLOSE, fd);
  }

  read(fd: number, buf: Uint8Array, count: number): SyscallResult<number> {
    return this.syscall(SyscallNumber.READ, fd, buf, count);
  }

  write(fd: number, buf: Uint8Array, count: number): SyscallResult<number> {
    return this.syscall(SyscallNumber.WRITE, fd, buf, count);
  }

  lseek(fd: number, offset: number, whence: number): SyscallResult<number> {
    return this.syscall(SyscallNumber.LSEEK, fd, offset, whence);
  }

  // Directory Operations
  chdir(path: string): SyscallResult<void> {
    return this.syscall(SyscallNumber.CHDIR, path);
  }

  getcwd(): SyscallResult<string> {
    return this.syscall(SyscallNumber.GETCWD);
  }

  mkdir(pathname: string, mode: number): SyscallResult<void> {
    return this.syscall(SyscallNumber.MKDIR, pathname, mode);
  }

  rmdir(pathname: string): SyscallResult<void> {
    return this.syscall(SyscallNumber.RMDIR, pathname);
  }

  // Time Operations
  time(): SyscallResult<number> {
    return this.syscall(SyscallNumber.TIME);
  }

  stime(t: number): SyscallResult<void> {
    return this.syscall(SyscallNumber.STIME, t);
  }

  // System Information
  uname(): SyscallResult<{
    sysname: string;
    nodename: string;
    release: string;
    version: string;
    machine: string;
  }> {
    return this.syscall(SyscallNumber.UNAME);
  }

  getuid(): SyscallResult<number> {
    return this.syscall(SyscallNumber.GETUID);
  }

  getgid(): SyscallResult<number> {
    return this.syscall(SyscallNumber.GETGID);
  }

  // Signal Operations
  signal(signum: number, handler: Function): SyscallResult<Function> {
    return this.syscall(SyscallNumber.SIGNAL, signum, handler);
  }

  // IPC Operations
  pipe(): SyscallResult<{ read: number; write: number }> {
    return this.syscall(SyscallNumber.PIPE);
  }
}

export const syscalls = new SystemCallInterface();

File: src/kernel/syscalls.c
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/fcntl.h>
#include <sys/times.h>
#include <sys/errno.h>
#include <sys/time.h>
#include <stdio.h>
#include <unistd.h>
#include <stddef.h>
#include "platform.h"

#undef errno
extern int errno;

// Environment variables
char *__env[1] = { 0 };
char **environ = __env;

// System call stubs for newlib
void _exit(int status) {
    (void)status;
    platform_boot_print("System exit called\n");
    for(;;);
}

int _close(int file) {
    (void)file;
    return -1;
}

int _execve(char *name, char **argv, char **env) {
    (void)name; (void)argv; (void)env;
    errno = ENOMEM;
    return -1;
}

int _fork() {
    errno = EAGAIN;
    return -1;
}

int _fstat(int file, struct stat *st) {
    (void)file;
    st->st_mode = S_IFCHR;
    return 0;
}

int _getpid() {
    return 1;
}

int _isatty(int file) {
    return (file == 0 || file == 1 || file == 2);
}

int _kill(int pid, int sig) {
    (void)pid; (void)sig;
    errno = EINVAL;
    return -1;
}

int _link(char *old, char *new) {
    (void)old; (void)new;
    errno = EMLINK;
    return -1;
}

int _lseek(int file, int ptr, int dir) {
    (void)file; (void)ptr; (void)dir;
    return 0;
}

int _open(const char *name, int flags, ...) {
    (void)name; (void)flags;
    return -1;
}

int _read(int file, char *ptr, int len) {
    (void)file; (void)ptr; (void)len;
    return 0;
}

caddr_t _sbrk(int incr) {
    // Simple heap management
    extern char _heap_start;
    static char *heap_end = NULL;
    char *prev_heap_end;
    
    if (heap_end == NULL) {
        heap_end = &_heap_start;
    }
    
    prev_heap_end = heap_end;
    heap_end += incr;
    
    return (caddr_t)prev_heap_end;
}

int _stat(const char *filepath, struct stat *st) {
    (void)filepath;
    st->st_mode = S_IFCHR;
    return 0;
}

clock_t _times(struct tms *buf) {
    (void)buf;
    return -1;
}

int _unlink(char *name) {
    (void)name;
    errno = ENOENT;
    return -1;
}

int _wait(int *status) {
    (void)status;
    errno = ECHILD;
    return -1;
}

int _write(int file, char *ptr, int len) {
    if (file == 1 || file == 2) { // stdout or stderr
        char buf[2] = { 0, 0 };
        for (int i = 0; i < len; i++) { buf[0] = ptr[i]; platform_boot_print(buf); }
        return len;
    }
    return -1;
}

// Time functions
int _gettimeofday(struct timeval *tv, struct timezone *tz) {
    (void)tz;
    if (tv) {
        tv->tv_sec = 0;
        tv->tv_usec = 0;
    }
    return 0;
}

// Non-underscore versions required by newlib
int gettimeofday(struct timeval *tv, void *tz) {
    return _gettimeofday(tv, (struct timezone *)tz);
}

int kill(int pid, int sig) {
    return _kill(pid, sig);
}

int getpid(void) {
    return _getpid();
}

void *sbrk(ptrdiff_t incr) {
    return _sbrk(incr);
}

int write(int fd, const void *buf, size_t count) {
    return _write(fd, (char *)buf, count);
}

int close(int fd) {
    return _close(fd);
}

off_t lseek(int fd, off_t offset, int whence) {
    return _lseek(fd, offset, whence);
}

int read(int fd, void *buf, size_t count) {
    return _read(fd, (char *)buf, count);
}

int fstat(int fd, struct stat *st) {
    return _fstat(fd, st);
}

int isatty(int fd) {
    return _isatty(fd);
}

// Init/fini functions required by newlib
void _init(void) {
    // Initialization code (empty for now)
}

void _fini(void) {
    // Finalization code (empty for now)
}
